### DISCUSSION TITLE
Fuzzy Matching | Does it work at all?

### DISCUSSION THREAD

**Discussion by bofrog** [DependencyTrack/dependency-track#2983] [General]
I have set up a test project in DependencyTrack using [bom.json.zip](https://github.com/DependencyTrack/dependency-track/files/12431682/bom.json.zip) as SBOM. It contains multiple components that are all intended to match OpenSSL (v1.0.2). Unfortunately, I could not observe ANY case where fuzzy matching would have an effect.

- For which use cases is the fuzzy matching feature intended for and what are its boundaries?
- Can somebody give me an example where fuzzy matching should work? (So that I can verify, if fuzzy matching is working at all)

*Note: I deployed DependencyTrack (v4.8.2) via Docker Compose exactly like it's described [here](https://docs.dependencytrack.org/getting-started/deploy-docker/#quickstart-docker-compose). I also activated all fuzzy matching options in the config panel (Analyzers -> Internal).*

Let me show you 3 cases where I'd have expected to see fuzzy matching

### Case 1: Only mandatory fields -> No match

![](https://bofrog.github.io/img/mandatory-only.png)

### Case 2: CPE with modified product field -> No match
`cpe:2.3:a:openssl:openssl_lib:1.0.2:*:*:*:*:*:*:*`

![](https://bofrog.github.io/img/name.png)

### Case 3: CPE with modified vendor field -> No match
`cpe:2.3:a:openssl_ltd:openssl:1.0.2:*:*:*:*:*:*:*`

![](https://bofrog.github.io/img/vendor.png)

### Sanity Check -> Matches
`cpe:2.3:a:openssl:openssl:1.0.2:*:*:*:*:*:*:*`

![](https://bofrog.github.io/img/sanity.png)
Results in
![](https://bofrog.github.io/img/sanity-r.png)

**Comment by officerNordberg** [comment]
Okay, I thought for sure I'd discover a massive logic hole based on the fact that your tests felt like they did a great job of exercising all sorts of aspects of the fuzzy matching which I will admit could be vastly improved by someone with greater Lucene knowledge but every unit test I wrote for every case you listed above passed.  My first suggestion is to rebuild your Vulnerable Software index.  This is at the heart of the fuzzy matching and I definitely ran into issues of it becoming corrupted or containing a subset of data in the database.  My _guess_ is the one match you hit on was a pure database hit rather than a Lucene search, but I could be mistaken.

![image](https://github.com/DependencyTrack/dependency-track/assets/901470/9935230e-bf32-4c80-9f0c-0667f984dbad)

**Reply by bofrog** [reply]
Thank you very much for your fast and detailed response!

I just executed the manual rebuild process you've pointed out. After this, I executed a portfolio update and an update on the project level. Unfortunately, the "Sanity Check" case is still the only one that is matching to vulnerabilities.

How did you conduct your test? Is it maybe a problem that I'm currently relying on the embedded H2 database?

---

On a side note: I executed the rebuild process on every single entity type that is listed in the settings (Projects, Components etc.). Interestingly, I couldn't check all types in one pass, because the server always skips after the first type if multiple ones are selected. The api server logs:

```
2023-08-25 10:34:10,902 INFO [ProjectIndexer] Starting reindex task. This may take some time.
2023-08-25 10:34:10,903 INFO [IndexManager] Deleting project index
2023-08-25 10:34:10,959 INFO [ProjectIndexer] Reindexing complete
2023-08-25 10:34:10,959 INFO [EventService] An singleton event (IndexEvent) was received but another singleton event of the same type is already in progress. Skipping.
```

This feels like a bug, because the UI does not reflect the fact that the process was canceled. Triggering one entity type at a time circumvents this behaviour. If this is a bug, does it also affect the periodic consistency check?

**Reply by bofrog** [reply]
Yay! I just managed to reproduce the state you've described (matching all but the "Mandatory Fields only" case). For this I had to create a new Project and uploaded the SBOM. In the new project fuzzy matching works as you've described **while in the old project the problem persists**.

Edit 1: To make things clearer. The "old project" consists of manually created components, so no SBOM upload was used. This seems to be the main difference as of now. Is it possible that such projects/components aren't affected by fuzzy matching?

**Reply by bofrog** [reply]
I conducted further tests to see DT's matching behaviour towards locally created vulnerabilities. I've observed some concerning cases.

*For testing purposes I've used a fictitious component **jobfuscator** by **acme** and created according vulnerabilities.*

SBOM: [bom_fuzzy_matching_2.zip](https://github.com/DependencyTrack/dependency-track/files/12472100/bom_fuzzy_matching_2.zip)

### Results

| SBOM Component         | Identifier                                          | Potenital match                                                                      | Actual match                                                                             |
|------------------------|-----------------------------------------------------|--------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------|
| jobfuscator (CPE)      | `cpe:2.3:a:acme:jobfuscator:2.1:*:*:*:*:*:*:*`       | `pkg:generic/acme/jobfuscator@2.1`<br>`cpe:2.3:a:acme:jobfuscator:2.1:*:*:*:*:*:*:*` | ❌ `pkg:generic/acme/jobfuscator@2.1`<br>✅ `cpe:2.3:a:acme:jobfuscator:2.1:*:*:*:*:*:*:*` |
| jobfuscator (purl)     | `pkg:generic/acme/jobfuscator@2.1`                  | `pkg:generic/acme/jobfuscator@2.1`<br>`cpe:2.3:a:acme:jobfuscator:2.1:*:*:*:*:*:*:*` | ✅ `pkg:generic/acme/jobfuscator@2.1`<br>✅ `cpe:2.3:a:acme:jobfuscator:2.1:*:*:*:*:*:*:*` |
| jobfuscator_lib (CPE)  | `cpe:2.3:a:acme:jobfuscator_lib:2.1:*:*:*:*:*:*:*`  | `pkg:generic/acme/jobfuscator@2.1`<br>`cpe:2.3:a:acme:jobfuscator:2.1:*:*:*:*:*:*:*` | ❌ no match                                                                               |
| jobfuscator_lib (purl) | `pkg:generic/acme/jobfuscator_lib@2.1`              | `pkg:generic/acme/jobfuscator@2.1`<br>`cpe:2.3:a:acme:jobfuscator:2.1:*:*:*:*:*:*:*` | ❌ no match                                                                               |
| openssl (CPE)          | `cpe:2.3:a:openssl:openssl:1.0.2:*:*:*:*:*:*:*`     | (NVD vulns)                                                                          | ✅ (NVD vulns)                                                                            |
| openssl (purl)         | `pkg:generic/openssl/openssl@1.0.2`                 | (NVD vulns)                                                                          | ✅ (NVD vulns)                                                                            |
| openssl_lib (CPE)      | `cpe:2.3:a:openssl:openssl_lib:1.0.2:*:*:*:*:*:*:*` | (NVD vulns)                                                                          | ✅ (NVD vulns)                                                                            |
| openssl_lib (purl)     | `pkg:generic/openssl/openssl_lib@1.0.2`             | (NVD vulns)                                                                          | ❌ no match                                                                               |

*Edit 1: Added further test cases*

**Comment by officerNordberg** [comment]
I also just tried this in a live instance of Dependency Track and was able to get 80~ vulnerabilities associated for all cases but the first, Mandatory Fields only.  Fuzzy matching relies on CPE but I'd have thought this should be a direct match.  There certainly is something not right, but otherwise my attempts at fuzzing the Vendor (openssl_ltd) and the Product (openssl_lib) appear to be working as expected or at least hoped.

**Reply by officerNordberg** [reply]
Ah Ha! If you specify only the mandatory fields I don't think you'll ever get a match using the Internal Analyzer.  
```
        String componentVersion;
        String componentUpdate;
        if (parsedCpe != null) {
            componentVersion = parsedCpe.getVersion();
            componentUpdate = parsedCpe.getUpdate();
        } else if (component.getPurl() != null) {
            componentVersion = component.getPurl().getVersion();
            componentUpdate = null;
        } else {
            // Catch cases where the CPE couldn't be parsed and no PURL exists.
            // Should be rare, but could lead to NPEs later.
            LOGGER.debug("Neither CPE nor PURL of component " + component.getUuid() + " provide a version - skipping analysis");
            return;
        }
        // In some cases, componentVersion may be null, such as when a Package URL does not have a version specified
        if (componentVersion == null) {
            return;
        }`

```
That `else` short circuits the versionRangeAnalysis when there's no PURL and no CPE.

**Reply by bofrog** [reply]
Thx for looking into root cause of this case. Should we create an issue or feature request out of this?

**Reply by officerNordberg** [reply]
np, I did create an issue directly from my comment above.  As for your reindexing issue, by all means.

Couple things to note, Portfolio Update does not do what you want it to do.  The only ways to get the InternalAnalyzer to run again is to click the Reanalyze button on a project (not refresh) or to update a component, changing values you've mentioned.  There's also no way to remove Vulnerabilities once you've "reanalyzed" and gotten too many due to fuzziness without uploading a new bom.

How I tested was I updated this embarrassingly utilitarian test to use your values https://github.com/DependencyTrack/dependency-track/blob/master/src/test/java/org/dependencytrack/search/FuzzyVulnerableSoftwareSearchManagerTest.java
I omitted my asserts to maintain a touch of brevity.  

```
    @Test
    public void fuzzyAnalysisLib() throws CpeParsingException, CpeValidationException {
        us.springett.parsers.cpe.Cpe justThePart = new us.springett.parsers.cpe.Cpe(Part.APPLICATION, "*", "*", "*", "*", "*", "*", "*", "*", "*", "*");
        // wildcard all components after part to constrain fuzzing to components of same type e.g. application, operating-system
        String fuzzyTerm = FuzzyVulnerableSoftwareSearchManager.getLuceneCpeRegexp(justThePart.toCpe23FS());
        SearchResult searchResult = toTest.searchIndex("product:openssl~0.88 AND " + fuzzyTerm);
        // Oddly validating lucene first cuz can't decouple from that.
        assertEquals(1, searchResult.getResults().size());
        assertEquals(1, searchResult.getResults().values().iterator().next().size());

        Component name_group = new Component();
        name_group.setName("OpenSSL");
        name_group.setGroup("OpenSSL");
        name_group.setVersion("1.0.2");
        Component lib_cpe = new Component();
        lib_cpe.setName("OpenSSL");
        lib_cpe.setGroup("OpenSSL");
        lib_cpe.setCpe("cpe:2.3:a:openssl:openssl_lib:1.0.2:*:*:*:*:*:*:*");

        Component vend_cpe = new Component();
        vend_cpe.setName("openssl");
        vend_cpe.setGroup("openssl");
        vend_cpe.setCpe("cpe:2.3:a:openssl_ltd:openssl:1.0.2:*:*:*:*:*:*:*");

        List<VulnerableSoftware> vs1 = toTest.fuzzyAnalysis(qm, name_group, null);
        name_group.setCpe("cpe:2.3:a:openssl:openssl:1.0.2:*:*:*:*:*:*:*");

        List<VulnerableSoftware> vs2 = toTest.fuzzyAnalysis(qm, lib_cpe, CpeParser.parse(lib_cpe.getCpe()));
        List<VulnerableSoftware> vs3 = toTest.fuzzyAnalysis(qm, vend_cpe, CpeParser.parse(vend_cpe.getCpe()));
        List<VulnerableSoftware> vs4 = toTest.fuzzyAnalysis(qm, vend_cpe, CpeParser.parse(name_group.getCpe()));
        List<VulnerableSoftware> vs5 = toTest.fuzzyAnalysis(qm, name_group, CpeParser.parse(vend_cpe.getCpe()));
        List<VulnerableSoftware> vs6 = toTest.fuzzyAnalysis(qm, name_group, CpeParser.parse(name_group.getCpe()));
}
    @BeforeClass
    public static void saveVsIndex() throws IOException {
        VulnerableSoftwareIndexer.getInstance().close();
        if (INDEX_TEMP_DIRECTORY.exists()) {
            FileUtils.deleteDirectory(INDEX_TEMP_DIRECTORY);
        }
        if (INDEX_DIRECTORY.exists()) {
            INDEX_DIRECTORY.renameTo(INDEX_TEMP_DIRECTORY);
        }
        VulnerableSoftware vs = new VulnerableSoftware();
        vs.setUuid(UUID.randomUUID());
        vs.setCpe23("cpe:2.3:a:openssl:openssl:1.0.2:*:*:*:*:*:*:*");
        vs.setProduct("openssl");
        VulnerableSoftwareIndexer.getInstance().add(vs);
        VulnerableSoftwareIndexer.getInstance().commit();
    }
```

**Reply by bofrog** [reply]
Thank you very much for your explanation. I wasn't aware of the importance of the "Reanalyze" button, thx for pointing that out! I think that the mechanics of the analysis process should be mentioned somewhere in the docs.

I will create an issue for the reindexing issue. And one about the fact that I need to reindex on a fresh instance in order to make fuzzy matching work at all.
